Extended workaround for handling symlinks in tar files

hhaensel opened the pull request:
I propose another workaround additional to `copyderef`.

`copyderef2` is automatically detected in case that neither `tempdir` nor `dest` supports symlinks.
In that case the symlinks of the tarfile are returned by the new function `list_tarball_symlinks`.
For this purpose the generic `list` functions have been modified to support a keyword argument `verbose`.
Regular expressions are used to parse the output and return a dictionary of symlinks.
The symlinks are exluded during extraction and the original source files of the symlinks are copied after extraction.

hhaensel commented:
I would be happy about people testing this on non-windows OS.
`copyderef2` can be forced by an environment variable [`BINARYPROVIDER_COPYDEREF2`](https://github.com/JuliaPackaging/BinaryProvider.jl/pull/163/files#diff-bc31fecc631ec0c23d226809fb26648bR637), which needs to be set to `true`.
(ImageMagick uses symlinks ;-) )

hhaensel commented:
It seems as if the verbose output of the tar listing on Mac-OS is different to that of a linux shell.
Therefore, my regex fails to identify the symlinks :-(
Tomorrow I will know, how a listing looks like and possibly adapt my regex...

hhaensel commented:
It seems I didn't have a proper example...

hhaensel commented:
Looks ok for me now :-)

hhaensel commented:
@staticfloat who is deciding on merging?
The regex should match most tar cases.

If a symlink starts with four numbers followed by a blank, it might be a problem ... 
One idea to make it more robust is to create one tar list example by
`echo a > a.txt && tar -cz a.txt | tar -tzv` and then determine the date pattern of the tar output...

Even more robust would be to first unzip, add a marker file to the tar and then list the content...
But we could leave this up to later, if this comes up as a problem.

hhaensel commented:
The last two changes were made to make the symlink parsing more robust as there are many formats around which might be even language dependent.

**Idea**
Check the listing of a demo file via `tar -c demoTar.txt | tar -tv`` and determine the number of parameters before the filename and finally construct a regex that filters the listing.

**Note**
gen_7z needs to contain the regex verbally, because splicing into functions has a lot of pitfalls (world-age problem)
For the tar section this is not a problem, because the regex is spliced into the tuple as an element and not as part of a function.

I consider this the last version for this PR and would recommend merging into master.

staticfloat commented on src/Prefix.jl:
```diff
@@ -470,6 +470,28 @@ function list_tarball_files(path::AbstractString; verbose::Bool = false)
+        error("Could not list contents of tarball $(tarball_path)")
+    end
+    output = collect_stdout(oc)
+    # @info("copyderef2: \r\n" * output)
```
Zap this comment

staticfloat commented on src/PlatformEngines.jl:
```diff
@@ -214,31 +214,46 @@ function probe_platform_engines!(;verbose::Bool = false)
+    gen_7z = (p) -> (unpack_7z(p), package_7z(p), list_7z(p), parse_7z_list, r"Path = ([^\r\n]+)\r?\n(?:[^\r\n]+\r?\n)+Symbolic Link = ([^\r\n]+)"s)
     compression_engines = Tuple[]
 
+    write("demoTar.txt","Demo file for tar listing")
```
This should be a temporary file created with `tempname()`; otherwise we are potentially overwriting user files.

staticfloat commented on src/PlatformEngines.jl:
```diff
@@ -214,31 +214,46 @@ function probe_platform_engines!(;verbose::Bool = false)
         package_tar = (in_path, tarball_path) ->
             `$tar_cmd -czvf $tarball_path -C $(in_path) .`
-        list_tar = (in_path) -> `$tar_cmd -tzf $in_path`
+        list_tar = (in_path; verbose = false) -> `$tar_cmd $(verbose ? "-tzvf" : "-tzf") $in_path`
```
Suggested change:

`$tar_cmd -tz$(verbose ? "v" : "")f`

hhaensel replied:
that was my first version, but it didn't work, because empty strings and strings containing blanks are added with quotation marks. Meanwhile, I found out that you have to use backticks :-)

hhaensel replied:
string interpolation in commands makes some very strange things, e.g.
```
julia> (x->`tar -tc$(x)f me`).(a)
4-element Array{Cmd,1}:
 `tar -tcvf me`
 `tar -tcvf me`
 `tar me`
 `tar -tcf me`

julia> (x->`tar $(x) me`).(a)
4-element Array{Cmd,1}:
 `tar v me`
 `tar v me`
 `tar me`
 `tar '' me`
```
I think, this is not intentional, is it?

staticfloat replied:
I don't think `Cmd`'s are made to be interpolated into other `Cmd`s; try only using `String`s:

```julia
julia> (v -> `tar -tz$(v ? "v" : "")f me`)(true)
`tar -tzvf me`

julia> (v -> `tar -tz$(v ? "v" : "")f me`)(false)
`tar -tzf me`
```

staticfloat commented on src/PlatformEngines.jl:
```diff
@@ -214,31 +214,46 @@ function probe_platform_engines!(;verbose::Bool = false)
+            # generic expression for symlink parsing
+            # this will fail, if the symlink contains space characters (which is highly improbable, though)
+            symlink_parser = r"^l.+? (\S+?)(?: -> (.+?))?\r?$"m
+        end
```
It would be very helpful to bake into this code, with comments, the different kinds of text the regexes are supposed to be matching.  E.g. capture the output across the different versions of `tar` that we are matching against, so we can see what the Regexes are doing.

staticfloat replied:
I see what you added, and that's great, but I literally meant examples of the text you're matching.  Something like:

```
# this regex matches the line "l. foo -> foo"
...
# this regex matches the line "l  foo -> foo"
```

Just so that we can see examples of what the lines look like.

staticfloat commented on src/PlatformEngines.jl:
```diff
@@ -214,21 +207,40 @@ function probe_platform_engines!(;verbose::Bool = false)
     compression_engines = Tuple[]
 
-    write("demoTar.txt","Demo file for tar listing")
+    (tmpfile, io) = mktemp()
```
More idiomatic:

```julia
mktemp() do (tmpfile, io)
	write(io, ...)
end
```

Also I imagine you want to get rid of the `@info(tmpfile)` below.

staticfloat replied:
This is also not a good place to use this, as we need to use `tmpfile` later.

staticfloat commented on src/PlatformEngines.jl:
```diff
@@ -648,34 +661,22 @@ function unpack(tarball_path::AbstractString, dest::AbstractString;
-            excludelist = joinpath(dest, "excludelist.txt")
-            write(excludelist, join([s[1] for s in symlinks], "\n"))
+            (excludelist, io) = mktemp()
+            write(io, join([s[1] for s in symlinks], "\n"))
```
Same thine here, use

```julia
mktemp() do (path, io)
	write(io, join([s[1] for s in symlinks], "\n"))
end
```

It's more fault-tolerant as it will catch exceptions that happen within `write()` and still close `io` when it breaks out.

hhaensel replied:
never used `do ... end` blocks until now, but it returns an error
```
MethodError: no method matching (::getfield(Main, Symbol("##19#21")))(::String, ::IOStream)
Closest candidates are:
mktemp(::getfield(Main, Symbol("##19#21")), ::String) at file.jl:555
mktemp(::Function) at file.jl:553
top-level scope at none:0
```

SimonDanisch replied:
I think it needs to be:
```Julia
mktemp() do path, io
	write(io, join([s[1] for s in symlinks], "\n"))
end
```
Since `(path, io)` does destructuring, which turns the function into a one arg function!

hhaensel replied:
I tested your proposal, but I need `tmpfile` later for `rm(tmpfile; force = true)`
So the version would look like
```
local tmpfile
mktemp() do path, io
	tmpfile = path
	write(io, "Demo file for tar listing (Julia package BinaryProvider.jl)")
	close(io)
end
```
Should I proceed with this version?

SimonDanisch replied:
I think the whole idea of the closure form is, that it deletes everything after the function is executed:
```
  mktemp(f::Function, parent=tempdir())

  Apply the function f to the result of mktemp(parent) and remove the
  temporary file upon completion.
```

SimonDanisch replied:
same goes for the `io`, so no need to close it!

hhaensel replied:
but I have somthing like

```
if copyderef  <create excludelist file> end
<unpack exempting exludelist>
if copyderef  <remove excludelist file> end
```
So I don't see how to use the do block.
Maybe, I am overlooking something?

SimonDanisch replied:
Maybe it's not the right choice here then ;)

staticfloat replied:
Simon is right; this is not the right choice in this case. :)  Sorry, I didn't fully realize all the places this was being used.  :)

staticfloat commented on src/PlatformEngines.jl:
```diff
@@ -239,7 +251,8 @@ function probe_platform_engines!(;verbose::Bool = false)
                 Jjz = "j"
             end
-            return `$tar_cmd -x$(Jjz)f $(tarball_path) --directory=$(out_path) $(excludelist=="" ? [] : ["--exclude-from=" * excludelist])`
+            excludeoption = excludelist=="" ? `` : `--exclude-from='$excludelist'`
```
Don't use backticks when interpolating into other backticks; just use Strings.

staticfloat commented on src/PlatformEngines.jl:
```diff
@@ -347,13 +434,14 @@ function probe_platform_engines!(;verbose::Bool = false)
+    for (test, unpack, package, list, parse, parse_symlinks) in compression_engines
         if probe_cmd(`$test`; verbose=verbose)
             # Set our compression command generators
             gen_unpack_cmd = unpack
```
Since you're changing the signature of `gen_unpack_cmd` here by adding an `excludelist` option, you should add that to the generic version up top that just errors out, so that we don't get a MethodError in the future.

staticfloat commented on src/PlatformEngines.jl:
```diff
@@ -193,18 +186,18 @@ function probe_platform_engines!(;verbose::Bool = false)
     # the correct 7z given the path to the executable:
     unpack_7z = (exe7z) -> begin
-        return (tarball_path, out_path) ->
+        return (tarball_path, out_path, excludelist = "") ->
```
Let's default `excludelist` to `nothing`

staticfloat commented on src/PlatformEngines.jl:
```diff
@@ -214,31 +207,125 @@ function probe_platform_engines!(;verbose::Bool = false)
+        end
         # Some tar's aren't smart enough to auto-guess decompression method. :(
-        unpack_tar = (tarball_path, out_path) -> begin
+        unpack_tar = (tarball_path, out_path, excludelist = "") -> begin
```
Let's default `excludelist` to `nothing`

staticfloat commented on src/PlatformEngines.jl:
```diff
@@ -630,25 +718,32 @@ Unpack tarball located at file `tarball_path` into directory `dest`.
+    # This is to work around filesystems that are mounted (such as SMBFS filesystems)
+    # that do not support symlinks.
+
+    excludelist = ""
```
default `excludelist` to `nothing`

hhaensel replied:
I digged a bit into interpolation from commands and I think I now understand the behavior:
```
x = ``
`command -o$x` # will evaluate to `command` and drop the option completely
x = "my file"
`command -o$x` # will evaluate to `command '-omy file'`
```
So what about defaulting ```excludelist::Union{AbstractString, Cmd} = `` ```
Then we can skip the `excludelist == nothing ? ... : ...` and simply use
```
        pipeline(`$exe7z x $(tarball_path) -y -so`,
                 `$exe7z x -si -y -ttar -o$(out_path)  -x@$(excludelist)`)
```
and
```
return `$tar_cmd -x$(Jjz)f $(tarball_path) --directory=$(out_path) --exclude-from=$(excludelist)`
```
I rewrote it and it workslike a charm ...

staticfloat replied:
So what's happening here is a kind of surprising interaction in the `Cmd` parser where it's seeing `command -o$x` as a cartesian product between arrays to be expanded, as described [in the docs here](https://docs.julialang.org/en/v1/manual/running-external-programs/index.html#command-interpolation-1).  Essentially, it looks at the `Cmd` like a container, and tries to iterate over it, but the empty `Cmd` is seen as a zero-dimensional object, and this disables all iteration, so we never see the first part (the `"-o"`).

This is not very intuitive at all, and I would not be surprised if this changes in a future version of Julia, so I think it's best not to rely on it.  Let's not use `Cmd` objects at all when interpolating into other `Cmd` objects and only use Strings; that's what is documented best, and is least likely to change in a future version.

Using `nothing` as the sentinel value for "do nothing" is pretty idiomatic, so I think thats better than using something else like a `Cmd` or an empty string; it fits in better with the rest of the APIs in Julia.

Thanks for putting so much work into this!

hhaensel commented:
I can easily agree to your comment, so I reverted my changes.

During testing I realised that Windows 10 has its own tar command which outputs \r\n.
Therefore, I have removed trailing `\r` in `parse_tar_list`, I hope that's ok.
Alternatively, the parse_tar_list could be done with a regex.
`lines = [m.captures[1] for m in eachmatch(r"^(.+?)(?<!/\r|/)\r?$"m, output)]`
For 7z this is also possible with
`r".+:\d{2}\s+\.\D+\d+\D+\d+\s+(?:\./|\.\)?(.+?)\r?$"m`

The tests at 603-605 of [runtests.jl](https://github.com/JuliaPackaging/BinaryProvider.jl/blob/master/test/runtests.jl#L603) don't pass if one choses `tar` or `busybox tar` in Windows, since
```
@test joinpath("bin", "bar.sh") in contents
```
fails due to different fileseparators. One do something like
```
joinpath("etc/qux.conf") in replace.(contents, ["\\" => "/"])
```
but I see that it's less elegant ;-)

So please comment, whether these two changes (regex and runtest.jl) should be made.
Otherwise, I hope that we are done with this PR...

staticfloat commented on src/PlatformEngines.jl:
```diff
@@ -19,7 +19,7 @@ gen_download_cmd = (url::AbstractString, out_path::AbstractString) ->
 
 """
-    gen_unpack_cmd(tarball_path::AbstractString, out_path::AbstractString; excludelist::Union{AbstractString, Cmd} = ``)
+    gen_unpack_cmd(tarball_path::AbstractString, out_path::AbstractString; excludelist::AbstractString = nothing)
```
`nothing` is not an `AbstractString`, so this won't work; you need to have `excludelist` be a `Union{AbstractString, Nothing}`.

staticfloat commented on src/PlatformEngines.jl:
```diff
@@ -29,7 +29,7 @@ This option is mainyl used to exclude symlinks from extraction (see: `copyderef`
 automatically called upon first import of `BinaryProvider`.
 """
-gen_unpack_cmd = (tarball_path::AbstractString, out_path::AbstractString; excludelist::Union{AbstractString, Cmd} = ``) ->
+gen_unpack_cmd = (tarball_path::AbstractString, out_path::AbstractString; excludelist::AbstractString = nothing) ->
```
Same here

hhaensel replied:
Stupid me! Thanks for your vigilance and patience - it's my first PR ;-)

staticfloat replied:
You're doing a great job!  Keep up the good work!

staticfloat commented:
I think the cleanest solution for the file path separation issue on Windows is to translate the filenames to be consistent across all platforms within `list_tarball_files()`.  So far, we've been using UNIX-y tools so this hasn't been a problem, but now that we have Windows-native tools, we need to check.  :P  So let's just have `list_tarball_files` always use `/` since that is valid on all platforms.  You can split the filenames by `"\\"` and then join them with `"/"` and return that.

hhaensel commented:
Oops, needs some rework, probably tomorrow ...

hhaensel commented:
The regex that I had unintentionally left in a modified version of tar_list should have been
`r"^(?:\./|\.\\)?(.+?)(?<!/|/\r)\r?$"m` ...
That would have worked. But the existing version is longer code but probably faster, so I left it unchanged except the removal of the `\r` line endings.

hhaensel commented:
I just realised that there is no mechanism for erroring out in case that `probe_platform_engines!()` has not been run.
What do you think about [adding](https://github.com/hhaensel/BinaryProvider.jl/blob/master/src/Prefix.jl#L479)

```
    if !isdefined(BinaryProvider, :gen_symlink_parser)
        error("Call `probe_platform_engines!()` before `list_tarball_symlinks()`")
    end
```

Btw, all error messages in the existing code still use `probe_platform_engines()` instead of `probe_platform_engines!()`
Shall I replace these within that PR?

Also, if you find my regex explanations too verbose, I might cut them down. Or feel free to redact them yourself to your own taste.

staticfloat commented on src/PlatformEngines.jl:
```diff
@@ -130,11 +132,6 @@ function probe_symlink_creation(dest::AbstractString)
 
-# Global variable that tells us whether tempdir() can have symlinks
-# created within it.
-tempdir_symlink_creation = false
```
Every time a `global` is removed from a file, an angel sheds a single tear of joy.

staticfloat commented:
I'm not certain why the tests are failing on Windows.  I may need to dig into that; it's possible this symlink thing isn't working for one of the test cases, but I don't have time to dig into it right now.  If you have access to a windows machine and can run the test cases locally, that might help.

hhaensel commented:
> 
> 
> I'm not certain why the tests are failing on Windows. I may need to dig into that; it's possible this symlink thing isn't working for one of the test cases, but I don't have time to dig into it right now. If you have access to a windows machine and can run the test cases locally, that might help.

Tests are failing only, if I manually chose tar to be the compression engine. Tar on windows behaves like a regular unix-y tar and lists its contents with regular slashes, whereas 7z uses backslashes.
When we would change that behavior but I can't judge, whether this would do something unexpected at a different place. In my symlinks workaround I need `/` for listing in `excludelist` but I use ``\` for the copying ...
Anyhow, I will now push the final version of this PR.

staticfloat commented:
Thanks @hhaensel!  I will correct the `/` vs. `\` problems in another PR.  This has been a long time coming, but I'm glad you stuck it through!

hhaensel commented:
I am really happy to see this merged - and I am a bit proud, as this is my first PR ever :-)
@staticfloat Thanks for your support! I learned a lot from the discussion we had.
